### PR TITLE
Update afterglow.py and kilonova.py

### PR DESCRIPTION
**Pull request by AkshayPanayada**

### CODE REVIEW

File: skysurvey/target/afterglow.py [hunk 1]
```diff
@@ -1,19 +1,11 @@
...
-    import afterglowpy
-except:
-    raise ImportError("could not import afterglowpy ; run pip install afterglowpy")
-
 
-from .core import Transient
-
-
-__all__ = ["Afterglow"]
+import afterglowpy
```
Comment (mcoughlin) [inline skysurvey/target/afterglow.py:5]: You should leave:

try:
    import afterglowpy
except:
    raise ImportError("could not import afterglowpy ; run pip install afterglowpy")

as before.

**Review by mcoughlin — CHANGES_REQUESTED** [review]
You should leave the afterglowpy import as optional.

File: skysurvey/target/afterglow.py [hunk 2]
```diff
@@ -47,16 +39,27 @@
 
 template = sncosmo.Model(sncosmo.TimeSeriesSource(phases, wave, np.array(flux)))
 
-class Afterglow( Transient ):
+def rate(redshift_):
+    psiarr=[]
+    for z in redshift_:
+        psi=0.0157+0.118*z/(1+(z/3.23)**4.66)
+        psiarr.append(psi)
+    return psiarr
```
Comment (mcoughlin) [inline skysurvey/target/afterglow.py:47]: I suggest:
return 0.0157+0.118*np.array(redshift_)/(1+(np.array(redshift_)/3.23)**4.66)

**Review by mcoughlin — CHANGES_REQUESTED** [review]
I think the rate function can be a lot simpler.

File: skysurvey/target/afterglow.py [hunk 3]
```diff
@@ -47,16 +43,27 @@
 
 template = sncosmo.Model(sncosmo.TimeSeriesSource(phases, wave, np.array(flux)))
 
-class Afterglow( Transient ):
```
Comment (mcoughlin) [inline skysurvey/target/afterglow.py:50]: This should still be the same Class, right?

File: skysurvey/target/afterglow.py [hunk 4]
```diff
@@ -1,19 +1,15 @@
...
 from scipy import stats
-
 import sncosmo
-try:
-    import afterglowpy
-except:
-    raise ImportError("could not import afterglowpy ; run pip install afterglowpy")
-
 
-from .core import Transient
```
Comment (mcoughlin) [inline skysurvey/target/afterglow.py:12]: This was correct already? Or not?

File: skysurvey/target/afterglow.py [hunk 5]
```diff
@@ -47,16 +40,28 @@
 
 template = sncosmo.Model(sncosmo.TimeSeriesSource(phases, wave, np.array(flux)))
```
Comment (JasonSev) [inline skysurvey/target/afterglow.py:41]: sncosmo.TimeSeriesSource requires `phases` to be in units of days, but it is currently given in seconds.  I suggest changing `phases` in this line to `phases/3600/24` or something similar

Comment (AkshayPanayada) [reply skysurvey/target/afterglow.py]: sure. I changed it and submitted a PR again

Comment (mcoughlin) [reply skysurvey/target/afterglow.py]: @AkshayPanayada Can you please copy in the current version of the file and make your changes on that to not change all of the whitespace? Additionally, please fix the rate code to as suggested in the comment above (i.e. return 0.0157+0.118*np.array(redshift_)/(1+(np.array(redshift_)/3.23)**4.66))

Comment (AkshayPanayada) [reply skysurvey/target/afterglow.py]: sure. Will do that

**Review by mcoughlin — APPROVED** [review]
LGTM

**Comment by mcoughlin** [comment]
@MickaelRigault Can you take a look and see about merging?

**Comment by MickaelRigault** [comment]
Hello guys, I updated Kilonova doing PR #20 to account for #19.

**Comment by mcoughlin** [comment]
Thanks @MickaelRigault ! @AkshayPanayada, I think you can close this one now.

**Comment by MickaelRigault** [comment]
Is the afterglow from main what you need?

**Comment by mcoughlin** [comment]
@MickaelRigault The main question is how to allow like E0 and other afterglow parameters vary: https://github.com/MickaelRigault/skysurvey/blob/main/skysurvey/target/afterglow.py#L27

**Comment by MickaelRigault** [comment]
I will look into that tomorrow.

Should all parameter be allowed to vary? (The variable names are not super explicit...)

**Comment by mcoughlin** [comment]
@MickaelRigault 

     'thetaObs':    0.05,   # Viewing angle in radians
     'E0':          1.0e53, # Isotropic-equivalent energy in erg
     'thetaCore':   0.1,    # Half-opening angle in radians
     'n0':          1.0,    # circumburst density in cm^{-3}
     'p':           2.2,    # electron energy distribution index
     'epsilon_e':   0.1,    # epsilon_e
     'epsilon_B':   0.01,   # epsilon_B
     'xi_N':        1.0,    # Fraction of electrons accelerated
     'd_L':         1.0e28, # Luminosity distance in cm
     'z':           0.55

I.e. everything but spectral type. Sensible ranges are:

https://github.com/nuclear-multimessenger-astronomy/nmma/blob/main/priors/TrPi2018.prior

**Comment by AkshayPanayada** [comment]
Should we mention about these ranges in the code?

**Comment by mcoughlin** [comment]
@AkshayPanayada I think they should be the default ranges sampled.

**Comment by MickaelRigault** [comment]
Looked into this. 

The issue is that your current implementation, the source is generated from a 2D flux-phase surface given your parameters. So none of them are  sncosmo.Source parameter. This could not work. 

The best would be to do a sncosmo.Source that depends on them, but redshift and t0 could not be source parameter (I think, to be confirmed) as this is captured by Model (model = sncosmo.Model(source))

something like:

```python
class AfterglowSource(sncosmo.Source):

    _param_names = ['amplitude', 
                    "thetaobs", "e0",
                    "thetacore", "n0",
                    "p","epsilon_e", "epsilon_b",
                    "xi_n"]
    param_names_latex = ['A', 'thetaObs', "E0", 
                        "thetaCore", "n0", "p",
                        "epsilon_e", "epsilon_B",
                        "xi_N"]   # used in plotting display

    def __init__(self, phase, wave, name=None, version=None):
        
        self.name = name
        self.version = version
        self._phase = phase
        self._wave = wave
        self._parameters = np.array([1., 
                                     0.05, 1.0e53, 0.1,
                                     1.0, 2.2, 0.1, 0.01, 
                                     1.0])  # initial parameters

    def _flux(self, phase, wave):
        # afterglowpy here
```

**Comment by AkshayPanayada** [comment]
So to be clear,
We should not use the afterglowpy package and instead use a sncosmo source
and modify the code accordingly .

I also noted an issue where the model could not be fitted properly while
generating a lightcurve. Will this issue be resolved by changing the model
to sncosmo one?

**Pull request closed without merging**